src/MLambda/Matrix.hs

```diff
@@ -99,35 +130,131 @@ MkNDArr a `cross` MkNDArr b = unsafePerformIO $ evalContT do
   , quoteType = error "Not implemented"
   , quoteDec = error "Not implemented"
   }
 
-matT :: forall e m n .
-  ( Storable e
-  , Read e
-  , KnownNat m
-  , KnownNat n
-  ) => String -> TH.CodeQ (NDArr [m, n] e)
-matT s = Code.do
-  let l = map (map (read @e) . words)
-        $ takeWhile (not . null)
-        $ dropWhile null
-        $ lines s
-  when (length l /= natVal m) $ fail "Wrong row count"
-  when (any ((/= natVal n) . length) l) $ fail "Wrong column count"
-  let dat = Storable.fromList $ concat l
+matE :: String -> TH.Q TH.Exp
+matE s = do
+  let expr prev =
+        TH.dyn <$> liftM2 (:) (P.satisfy isAlpha) (P.munch isAlphaNum)
+          <|>
+        do
+          str <- P.manyTill P.get $ P.char ')'
+          case Haskell.parseExp (prev ++ str ++ ")") of
+            Right e -> pure $ pure e
+            Left _  -> expr (prev ++ str ++ ")")
+  let item = Right <$> P.readS_to_P (reads @Double)
+         <|> Left <$> (P.char '$' *> expr "")
+  let space = P.skipMany $ P.char ' '
+  let line = space *>P.endBy1 item space
+  let matrix = P.skipSpaces *> P.sepBy1 line (P.char '\n') <* P.skipSpaces <* P.eof
+  l <- maybe (fail "Failed parse") (pure . fst) $ listToMaybe $ P.readP_to_S matrix s
+  let m = length l
+  n <- maybe (fail "Empty matrix specified") (pure . length) $ listToMaybe l
+  when (n == 0) $ fail "Empty matrix specified"
+  when (any ((/= n) . length) l) $ fail "Wrong column count"
+  let indices = concat $ zipWith (<*>) (map ((:[]) . (+)) [0, n..]) $ map (findIndices isLeft) l
+  let exps = map (fromLeft (TH.dyn "bruh")) $ concatMap (filter isLeft) l
+  let l' = map (fromRight 0) $ concat l
+  let dat = Storable.fromList l'
       (fptr, offset, len) =
         Storable.unsafeToForeignPtr $ Storable.unsafeCast dat
-      bytes = TH.unsafeCodeCoerce
-            $ TH.litE $ TH.bytesPrimL
+
+  let bytes = TH.litE $ TH.bytesPrimL
             $ TH.mkBytes fptr (fromIntegral offset)
             $ fromIntegral len
-      size = natVal m * natVal n
-  [||
-    unsafePerformIO do
-      ptr <- newForeignPtr_ $ Ptr $$bytes
-      pure $ MkNDArr $ Storable.unsafeFromForeignPtr0 ptr size
-    ||]
+      size = m * n
+      tm = TH.litT $ TH.numTyLit $ toInteger m
+      tn = TH.litT $ TH.numTyLit $ toInteger n
+  let raw = [|
+        unsafeDupablePerformIO do
+          ptr <- newForeignPtr_ $ Ptr $bytes
+          pure $ Storable.unsafeFromForeignPtr0 ptr size
+        |]
+  if null indices
+    then [| MkNDArr @'[$tm, $tn] @Double $raw |]
+    else [| runST do
+              mvec <- Storable.thaw $raw
+              $(TH.doE $ flip map (zip indices exps)
+                \(i, e) -> TH.noBindS [| Mutable.write mvec i $e |])
+              vec <- Storable.unsafeFreeze mvec
+              pure $ MkNDArr @'[$tm, $tn] @Double vec
+      |]
+
+-- | `Index dim` is the type of indices of an array of type `NDArr dim e`.
+data Index (dim :: [Natural]) where
+  (:.) :: Index '[n] -> Index d -> Index (n : d)
```

Comment:
I think you can try `d:ds` here instead of just `d` for GHC to recognize that either of `:.` and `I` are `COMPLETE` and get rid of the pragmas